File: frontend/src/lib/api/mappings.ts
import { apiGet, apiSend } from "@/lib/api/client";
import type { ProfileColumn } from "@/lib/api/datasets";

export type CanonicalField = { id: string; code: string; domain: string; name: string; description: string; expected_data_type: string; expected_grain: string; unit_type: string; entity_type: string; aliases: string[]; active: boolean };
export type MappingStatus = "SUGGESTED" | "CONFIRMED" | "REJECTED" | "UNMAPPED" | "NEEDS_REVIEW";
export type Mapping = { id: string; dataset_version_id: string; source_column_id: string; status: MappingStatus; confidence: number | null; confidence_level: "HIGH" | "MEDIUM" | "LOW" | null; mapping_method: string; evidence: { kind: string; message: string }[]; alternatives: { canonical_field_id: string; code: string; score: number }[]; confirmed_by_user: boolean; grain: string; source_column: ProfileColumn; canonical_field: CanonicalField | null };
export type MappingCollection = { dataset_id: string; dataset_version_id: string; mappings: Mapping[]; summary: { total: number; confirmed: number; needs_review: number; suggested: number; rejected: number; unmapped: number; domains: { domain: string; confirmed_fields: number; catalog_fields: number; status: "PARTIAL" | "MISSING" }[] } };

export const getCanonicalFields = () => apiGet<CanonicalField[]>("/api/v1/canonical-fields");
export const getMappings = (datasetId: string) => apiGet<MappingCollection>(`/api/v1/datasets/${datasetId}/mappings`);
export const suggestMappings = (datasetId: string) => apiSend<MappingCollection>(`/api/v1/datasets/${datasetId}/mappings/suggest`, { method: "POST", body: JSON.stringify({}) });
export const updateMapping = (mappingId: string, status: MappingStatus, canonicalFieldId?: string | null) => apiSend<Mapping>(`/api/v1/mappings/${mappingId}`, { method: "PATCH", body: JSON.stringify({ status, canonical_field_id: canonicalFieldId ?? null }) });
export const confirmMapping = (mappingId: string, canonicalFieldId?: string | null) => apiSend<Mapping>(`/api/v1/mappings/${mappingId}/confirm`, { method: "POST", body: JSON.stringify({ canonical_field_id: canonicalFieldId ?? null }) });
export const rejectMapping = (mappingId: string) => apiSend<Mapping>(`/api/v1/mappings/${mappingId}/reject`, { method: "POST", body: JSON.stringify({}) });

File: backend/app/schemas/mappings.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict

from app.schemas.datasets import ColumnRead

MappingStatus = Literal["SUGGESTED", "CONFIRMED", "REJECTED", "UNMAPPED", "NEEDS_REVIEW"]


class CanonicalFieldRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    code: str
    domain: str
    name: str
    description: str
    expected_data_type: str
    expected_grain: str
    unit_type: str
    entity_type: str
    aliases: list[str]
    active: bool


class MappingRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    dataset_version_id: str
    source_column_id: str
    status: str
    confidence: float | None
    confidence_level: str | None
    mapping_method: str
    evidence: list[dict[str, str]]
    alternatives: list[dict[str, Any]]
    confirmed_by_user: bool
    grain: str
    created_at: datetime
    updated_at: datetime
    source_column: ColumnRead
    canonical_field: CanonicalFieldRead | None


class MappingUpdate(BaseModel):
    status: MappingStatus
    canonical_field_id: str | None = None


class MappingChoice(BaseModel):
    canonical_field_id: str | None = None


class DomainCoverage(BaseModel):
    domain: str
    confirmed_fields: int
    catalog_fields: int
    status: Literal["PARTIAL", "MISSING"]


class MappingSummary(BaseModel):
    total: int
    confirmed: int
    needs_review: int
    suggested: int
    rejected: int
    unmapped: int
    domains: list[DomainCoverage]


class MappingCollection(BaseModel):
    dataset_id: str
    dataset_version_id: str
    mappings: list[MappingRead]
    summary: MappingSummary
